docs/java/how-to-develop-a-workflow-definition-in-java.md

```diff
@@ -0,0 +1,186 @@
+To illustrate this, lets' say that we define the following **invalid** code:
+
+```java
+// INVALID CODE!
+public interface BaseWorkflow {
+    @WorkflowMethod
+    void retryNow();
+}
+
+@WorkflowInterface
+public interface Workflow1 extends BaseWorkflow {}
+
+@WorkflowInterface
+public interface Workflow2 extends BaseWorkflow {}
+```
+
+If we attempt to register implementations of Workflow1 and Workflow2 with a Worker, it will fail. Let's say that we have:
+
+```java
+worker.registerWorkflowImplementationTypes(
+        Workflow1Impl.class, Workflow2Impl.class);
+```
+
+This registration will fail with:
+
+```text
+java.lang.IllegalStateException: BaseWorkflow workflow type is already registered with the worker
+```
+
+## Implementing Workflows
+
+A Workflow implementation implements a Workflow interface. Each time a new Workflow Execution is started,
+a new instance of the Workflow implementation object is created.
+Then, one of the methods (depending on which Workflow type has been started) annotated with `@WorkflowMethod` can be invoked.
+As soon as this method returns, the Workflow execution is considered as completed.
+
+Workflow methods annotated with `@QueryMethod` and `@SignalMethod` can be invoked during a Workflow's execution.
+
+Note that methods annotated with `@QueryMethod` can be invoked even when a Workflow is in the `Completed`
+state.
+
+### Workflow Implementation Constraints
+
+Temporal uses the [Event Sourcing pattern](https://docs.microsoft.com/en-us/azure/architecture/patterns/event-sourcing) to recover
+the state of a Workflow object including its threads and local variable values.
+In essence, every time a Workflow state has to be restored, its code is re-executed from the beginning.
+Note that during replay, successfully executed Activities are not re-executed as their results are already recorded
+in the Workflow event history.
+
+The following constraints apply when writing Workflows:
+
+- Do not use constructs that rely on system time.
+- Do not use any mutable global variables in your Workflow implementations. This will assure that multiple Workflow instances are fully isolated.
+- Do not call any non-deterministic functions like non-seeded random or `UUID.randomUUID()` directly from the Workflow code. The Temporal SDK provides specific API for calling non-deterministic code in your Workflows.
+- Do not use any programming language constructs that rely on system time. For example, only use `Workflow.currentTimeMillis()` to get the current time inside a Workflow.
+- Do not use native Java `Thread` or any other multi-threaded classes like `ThreadPoolExecutor`. Use `Async.function` or `Async.procedure`,
+  provided by the Temporal SDK, to execute code asynchronously.
+- Do not use any synchronization, locks, and other standard Java blocking concurrency-related classes besides those provided
+  by the Workflow class. There is no need for explicit synchronization because multi-threaded code inside a Workflow is
```

Comment:
```suggestion
  There is no need for explicit synchronization because multi-threaded code inside a Workflow is executed one thread at a time and under a global lock.
```